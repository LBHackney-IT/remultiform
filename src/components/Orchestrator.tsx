import PropTypes from "prop-types";
import React from "react";

import { DatabaseContext } from "../helpers/DatabaseContext";
import { PageComponentWrapper } from "../helpers/PageComponentWrapper/PageComponentWrapper";

import { NamedSchema, Schema } from "../store/types";

import { DatabaseProvider } from "./DatabaseProvider";
import { Page } from "./Page";

/**
 * A step in a multipage form.
 *
 * This represents a single step in the flow of the multipage form.
 */
export interface Step<
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  DBSchema extends NamedSchema<string, number, Schema> = any
> {
  /**
   * The key to uniquely identify the step
   *
   * This is referenced by {@link OrchestratorProps.currentStepKey}.
   */
  key: React.Key;

  /**
   * An ordered array of wrapped components to include in this step.
   *
   * Create {@link PageComponentWrapper|PageComponentWrappers} with
   * {@link PageComponentWrapper.wrapStatic} and
   * {@link PageComponentWrapper.wrapDynamic}.
   */
  componentWrappers: PageComponentWrapper<DBSchema>[];
}

/**
 * The proptypes for {@link Orchestrator}.
 */
export interface OrchestratorProps<
  DBSchema extends NamedSchema<string, number, Schema>
> {
  /**
   * A context wrapper for a {@link Database} instance.
   *
   * You must provide this if any of your components are instances of
   * {@link DynamicPageComponent} for them to work as expected.
   */
  context?: DatabaseContext<DBSchema> | null;

  /**
   * The key of the current step to be rendered.
   */
  currentStepKey?: React.Key | null;

  /**
   * An array of steps to possibly render based on
   * {@link OrchestratorProps.currentStepKey}.
   */
  steps: Step<DBSchema>[];
}

/**
 * A component for orchestrating the rendering of pages for a multipage form.
 */
export const Orchestrator = <
  DBSchema extends NamedSchema<string, number, Schema>
>(
  props: OrchestratorProps<DBSchema>
): JSX.Element => {
  const { context, currentStepKey, steps } = props;

  const currentStep =
    steps.find(({ key }) => key === currentStepKey) || steps[0];

  if (context) {
    return (
      <DatabaseProvider context={context}>
        <Page
          context={context}
          key={currentStep.key}
          componentWrappers={currentStep.componentWrappers}
        />
      </DatabaseProvider>
    );
  }

  return (
    <Page
      key={currentStep.key}
      componentWrappers={currentStep.componentWrappers}
    />
  );
};

Orchestrator.propTypes = {
  currentStepKey: PropTypes.string,
  steps: PropTypes.arrayOf(
    PropTypes.exact({
      key: PropTypes.oneOfType([
        PropTypes.string.isRequired,
        PropTypes.number.isRequired
      ]).isRequired,
      componentWrappers: PropTypes.arrayOf(
        PageComponentWrapper.propType.isRequired
      ).isRequired
    }).isRequired
  ).isRequired
} as PropTypes.ValidationMap<
  OrchestratorProps<NamedSchema<string, number, Schema>>
>;
