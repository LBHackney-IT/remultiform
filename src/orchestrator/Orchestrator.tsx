import PropTypes from "prop-types";
import React from "react";

import { NamedSchema, Schema, StoreNames } from "../database/types";
import { DatabaseContext } from "../database-context/DatabaseContext";
import { DatabaseProvider } from "../database-context/DatabaseProvider";
import { Step } from "../step/internal/Step";
import { StepDefinition, stepPropType } from "../step/StepDefinition";

/**
 * The proptypes for {@link Orchestrator}.
 */
export interface OrchestratorProps<
  DBSchema extends NamedSchema<string, number, Schema>,
  StoreName extends StoreNames<DBSchema["schema"]>
> {
  /**
   * A context wrapper for a {@link Database} instance.
   *
   * You must provide this if any of your components are instances of
   * {@link DynamicComponent} for them to work as expected.
   */
  context?: DatabaseContext<DBSchema> | null;

  /**
   * An array of steps to render.
   */
  steps: StepDefinition<DBSchema, StoreName>[];

  /**
   * The {@link StepDefinition.slug} to start the form on.
   *
   * Defaults to the slug of the first {@link StepDefinition} in
   * {@link OrchestratorProps.steps}.
   */
  initialSlug?: string | null;

  /**
   * Whether the {@link Orchestrator} should manage {@link StepDefinition} transitions or
   * you want to manage them yourself via routing, for example.
   *
   * Defaults to `true`.
   */
  manageStepTransitions?: boolean | null;

  /**
   * Whether the {@link Orchestrator} should wrap the steps in a
   * {@link DatabaseProvider} for you or you want to do that yourself.
   *
   * You might want to disable this if you have pages that exist outside of the
   * {@link Orchestrator}.
   *
   * Defaults to `true`.
   */
  provideDatabase?: boolean | null;

  /**
   * The callback to call when the {@link Orchestrator} is about to transition
   * to the next {@link StepDefinition}.
   *
   * This is called after the step has been submitted, and the
   * {@link Database} transaction has completed. It is called immediately before
   * it renders the new {@link StepDefinition} (if
   * {@link OrchestratorProps.manageStepTransitions} is `true`).
   *
   * Use this for any side effects that you might want when transitioning
   * {@link StepDefinition|steps} and for handling transitions to a
   * {@link StepDefinition.nextSlug} that isn't managed by the
   * {@link Orchestrator}.
   *
   * @param slug - The next {@link StepDefinition.slug}.
   */
  onNextStep?: ((slug?: string) => void) | null;
}

interface OrchestratorState {
  currentSlug?: string;
  nextSlug?: string;
}

/**
 * A component for orchestrating the rendering of
 * {@link StepDefinition|StepDefinitions} for a multipage form.
 */
export class Orchestrator<
  DBSchema extends NamedSchema<string, number, Schema>,
  StoreName extends StoreNames<DBSchema["schema"]>
> extends React.Component<
  OrchestratorProps<DBSchema, StoreName>,
  OrchestratorState,
  never
> {
  static propTypes: PropTypes.ValidationMap<
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    OrchestratorProps<NamedSchema<string, number, any>, string>
  > = {
    context: PropTypes.instanceOf(DatabaseContext),
    steps: PropTypes.arrayOf(stepPropType.isRequired).isRequired,
    initialSlug: PropTypes.string,
    manageStepTransitions: PropTypes.bool,
    onNextStep: PropTypes.func
  };

  static defaultProps: Partial<
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    OrchestratorProps<NamedSchema<string, number, any>, string>
  > = {
    manageStepTransitions: true,
    provideDatabase: true
  };

  state: OrchestratorState = {};

  /**
   * @ignore
   */
  render(): JSX.Element {
    const { context, provideDatabase } = this.props;

    const { slug, nextSlug, componentWrappers, submit } = this.currentStep();

    const step = (
      <Step
        key={slug}
        context={context}
        componentWrappers={componentWrappers}
        submit={submit}
        afterSubmit={(): void => {
          this.handleSubmit();
        }}
        nextSlug={nextSlug}
        onNextSlugChange={(slug?: string): void => {
          this.handleNextSlugChange(slug);
        }}
      />
    );

    if (provideDatabase && context) {
      return <DatabaseProvider context={context}>{step}</DatabaseProvider>;
    }

    return step;
  }

  private currentStep(): StepDefinition<DBSchema, StoreName> {
    const { steps, initialSlug, manageStepTransitions } = this.props;

    let currentSlug = initialSlug;

    if (manageStepTransitions && this.state.currentSlug) {
      currentSlug = this.state.currentSlug;
    }

    if (!currentSlug) {
      return steps[0];
    }

    const step = steps.find(({ slug }) => slug === currentSlug);

    if (!step) {
      throw new Error(`Step with slug "${currentSlug}" could not be found`);
    }

    return step;
  }

  private handleNextSlugChange(slug?: string): void {
    this.setState(state => ({ ...state, nextSlug: slug }));
  }

  private handleSubmit(): void {
    const { steps, manageStepTransitions, onNextStep } = this.props;
    const { nextSlug } = this.state;

    if (onNextStep) {
      onNextStep(nextSlug);
    }

    this.setState(state => ({ ...state, nextSlug: undefined }));

    if (!manageStepTransitions) {
      return;
    }

    if (nextSlug && steps.map(step => step.slug).includes(nextSlug)) {
      this.setState(state => ({ ...state, currentSlug: nextSlug }));
    } else if (!onNextStep) {
      throw new Error(
        `Unable to transition to next slug "${nextSlug}" due to it not ` +
          "existing in the managed steps, and there is no onNextStep prop to " +
          "call instead"
      );
    }
  }
}
