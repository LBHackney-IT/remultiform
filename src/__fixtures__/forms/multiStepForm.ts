import { Step } from "../../components/Orchestrator";

import { PageComponentWrapper } from "../../helpers/PageComponentWrapper/PageComponentWrapper";
import { StaticPageComponent } from "../../helpers/PageComponentWrapper/StaticPageComponent";

import { TestClassComponent } from "../components/TestClassComponent";
import { TestFunctionComponent } from "../components/TestFunctionComponent";

export const multiStepForm: { steps: Step[] } = {
  steps: [
    {
      key: "test-step-1",
      componentWrappers: [
        PageComponentWrapper.wrapStatic(
          new StaticPageComponent({
            key: "test-div",
            Component: "div",
            props: {}
          })
        ),
        PageComponentWrapper.wrapStatic(
          new StaticPageComponent({
            key: "test-class",
            Component: TestClassComponent,
            props: {
              content: "test class content"
            }
          })
        )
      ]
    },
    {
      key: "test-step-2",
      componentWrappers: [
        PageComponentWrapper.wrapStatic(
          new StaticPageComponent({
            key: "test-img",
            Component: "img",
            props: {
              src: "test.png"
            }
          })
        ),
        PageComponentWrapper.wrapStatic(
          new StaticPageComponent({
            key: "test-function",
            Component: TestFunctionComponent,
            props: {
              content: "test function content"
            }
          })
        )
      ]
    }
  ]
};
