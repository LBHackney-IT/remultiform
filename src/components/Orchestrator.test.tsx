import React from "react";
import { create } from "react-test-renderer";

import multiStepForm from "../__fixtures__/forms/multiStepForm";

import Orchestrator from "./Orchestrator";

it("renders correctly with all props", () => {
  const component = create(
    <Orchestrator currentStepKey="test-step-2" steps={multiStepForm.steps} />
  );

  expect(component).toMatchSnapshot();
});

it("renders correctly without `currentStepKey`", () => {
  const component = create(<Orchestrator steps={multiStepForm.steps} />);

  expect(component).toMatchSnapshot();
});
