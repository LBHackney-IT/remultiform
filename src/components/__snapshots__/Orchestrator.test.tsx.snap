// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with all props 1`] = `
Array [
  <img
    src="test.png"
  />,
  <div>
    test function content
  </div>,
]
`;

exports[`renders correctly without optional props 1`] = `
Array [
  <div />,
  <div>
    test class content
  </div>,
]
`;
